department-of-veterans-affairs/caseflow
# Video hearing days should not respect spreadsheet slot_count value

Resolves [CASEFLOW-1610](https://vajira.max.gov/browse/CASEFLOW-1610)

### Description
When uploading a spreadsheet with video hearing days, the video hearing days should have 12 slots regardless of what is in the "Number of Time Slots" column. That column is intended only for virtual hearing days.

### Acceptance Criteria
- [x] Building a schedule creates virtual days with the number of slots in the "Number of Time Slots" column in the spreadsheet.
- [x] Building a schedule creates video days with 12 slots, video days should ignore the "Number of Time Slots" column.

### Testing Plan
1. Sign in as user BVASYELLOW http://localhost:3000/test/users
2. Navigate to the build hearings schedule: http://localhost:3000/hearings/schedule/build
3. Click the "Upload Files" button
4. Fill out the start date as `04/01/2021` and the end date as `6/30/2021`
5. Upload the 'virtual-test-days.xlsx' spreadsheet from JIRA CASEFLOW-1610 ([direct link to spreadsheet](https://vajira.max.gov/secure/attachment/157519/virtual-test-days.xlsx))
6. Click continue to schedule the assignments
7. Confirm the time slot details
    - All video days should have no slot configuration displayed (number of slots, length of slots, first slot time).
    - Virtual days should have as many time slots as the "Number of Time Slots" column in the spreadsheet says.
8. **I recommend doing as much testing as possible before these next steps, since you have to `make reset` your db after these steps for the upload to work again.**
9. Click confirm assignments and confirm when the dialog pops up
8. Ensure that you get a success message indicating the hearing days were created
9. Run the following command to verify that only non-virtual days were assigned timeslot configuration values
```ruby
# All hearing days created today, output every unique combination of the relevant fields
HearingDay.all { |hd| hd.created_at.date == Date.today }.pluck(
  :request_type, :first_slot_time, :number_of_slots, :slot_length_minutes
  ).uniq

# Results like this show that there are no non-virtual days with slot configuration values
=> [["C", nil, nil, nil],
 ["R", nil, nil, nil],
 ["V", nil, nil, nil],
 ["R", "09:15", 8, 60],
 ["R", "8:30", 8, 60],
 ["R", "12:30", 5, 60],
 ["R", "13:30", 5, 60],
 ["R", "10:30", 7, 60],
 ["R", "9:30", 7, 60],
 ["R", "20:30", 5, 60]]

# Also useful for isolating the HearingDays you just created (thanks Rubaiyat)
HearingDay.last(100).select { |h| h.request_type == "V" }.map { |h| [h.number_of_slots, h.slot_length_minutes, h.first_slot_time, h.created_at] }

# We overload some db values, this deals with that
HearingDay.last(100).select { |h| h.request_type == "C" }.map { |h| [h[:number_of_slots], h[:slot_length_minutes], h[:first_slot_time], h.created_at] }
```
